src/crypto.rs

```diff
@@ -381,20 +369,152 @@ impl PrivateKey {
         let private_key_bytes = &key[..ED25519_PRIVATE_KEY_LENGTH];
         let public_key_bytes = &key[ED25519_PUBLIC_KEY_LENGTH..];
 
-        let key = Ed25519KeyPair::from_seed_and_public_key(private_key_bytes, public_key_bytes)
+        let private = Ed25519KeyPair::from_seed_and_public_key(private_key_bytes, public_key_bytes)
             .map_err(|err| Error::Encoding(err.to_string()))?;
 
         let public = PublicKey::new(
             KeyType::Ed25519,
             SignatureScheme::Ed25519,
             keyid_hash_algorithms,
-            key.public_key().as_ref().to_vec(),
+            private.public_key().as_ref().to_vec(),
+        )?;
+
+        Ok(Ed25519PrivateKey { private, public })
+    }
+}
+
+impl PrivateKey for Ed25519PrivateKey {
+    fn sign(&self, msg: &[u8]) -> Result<Signature> {
+        match &self.public.scheme {
+            SignatureScheme::Ed25519 => (),
+            s => Err(Error::IllegalArgument(format!(
+                "Key {:?} can't be used with scheme {:?}",
+                self.private, s
+            )))?,
+        }
+
+        let value = SignatureValue(self.private.sign(msg).as_ref().into());
+        Ok(Signature {
+            key_id: self.public.key_id().clone(),
+            value,
+        })
+    }
+
+    fn public(&self) -> &PublicKey {
+        &self.public
+    }
+}
+
+/// A structure containing information about an Rsa private key.
+pub struct RsaPrivateKey {
+    private: Arc<RsaKeyPair>,
+    public: PublicKey,
+}
+
+impl RsaPrivateKey {
+    /// Generate RSA key bytes in pkcs8 format.
+    ///
+    /// Note: `openssl` needs to the on the `$PATH`.
+    pub fn keypair() -> Result<Vec<u8>> {
+        let gen = Command::new("openssl")
+            .args(&[
+                "genpkey",
+                "-algorithm",
+                "RSA",
+                "-pkeyopt",
+                "rsa_keygen_bits:4096",
+                "-pkeyopt",
+                "rsa_keygen_pubexp:65537",
+                "-outform",
+                "der",
+            ])
+            .output()?;
+
+        let mut pk8 = Command::new("openssl")
+            .args(&[
+                "pkcs8", "-inform", "der", "-topk8", "-nocrypt", "-outform", "der",
+            ])
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        match pk8.stdin {
+            Some(ref mut stdin) => stdin.write_all(&gen.stdout)?,
+            None => return Err(Error::Opaque("openssl has no stdin".into())),
+        };
+
+        Ok(pk8.wait_with_output()?.stdout)
+    }
+
+    fn from_pkcs8(der_key: &[u8], scheme: SignatureScheme) -> Result<Self> {
+        if let SignatureScheme::Ed25519 = scheme {
```

Comment:
to hedge against future changes, instead I'd recommend:

```
match scheme {
    SignatureScheme:: RsaSsaPssSha256 | SignatureScheme:: RsaSsaPssSha512 => {}
    _ => { return Err(Error::IllegalArgument(format!("RSA keys do not support the signing scheme {}", scheme))); }
}
```

Or as an alternative, we could change `SigningScheme` to be:

```
#[non_exhaustive]
enum SigningScheme {
    Ed25519,
    Rsa(RsaSigningScheme),
    ...
}

#[non_exhaustive]
enum RsaSigningScheme {
    RsaSsaPssSha256,
    RsaSsaPssSha512,
}
```

Then we could change this function to take an `RsaSigningScheme` and never have to worry about the wrong scheme being passed in. The only catch is we'd have to make sure that serializing/deserializing a `SigningScheme` should produce a flat string.

Reply:
Did the simpler option.